basic_nodes/src/main/java/club/mondaylunch/gatos/basicnodes/ListHeadTailNodeType.java

```diff
@@ -0,0 +1,93 @@
+    private static final DataType<ExtractionMode> EXTRACTION_MODE_TYPE = DataType.register("extraction_mode_type", ExtractionMode.class);
+    @Override
+    public Map<String, DataBox<?>> settings() {
+        return Map.of(
+            "extraction_mode", getHeadExtractionBox()
+        );
+    }
+
+    @Override
+    public Set<NodeConnector.Input<?>> inputs(UUID nodeId, Map<String, DataBox<?>> settings, Map<String, DataType<?>> inputTypes) {
+        return Set.of(
+            new NodeConnector.Input<>(nodeId, "input", ListDataType.GENERIC_LIST)
+        );
+    }
+
+    @Override
+    public Set<NodeConnector.Output<?>> outputs(UUID nodeId, Map<String, DataBox<?>> settings, Map<String, DataType<?>> inputTypes) {
+        var outType = inputTypes.isEmpty() ? DataType.ANY : inputTypes.get("input");
+        return Set.of(
+            new NodeConnector.Output<>(nodeId, "first", outType),
+            new NodeConnector.Output<>(nodeId, "rest", outType.listOf())
+        );
+    }
+
+    @Override
+    @SuppressWarnings("unchecked")
+    public Map<String, CompletableFuture<DataBox<?>>> compute(Map<String, DataBox<?>> inputs, Map<String, DataBox<?>> settings, Map<String, DataType<?>> inputTypes) {
+        var inputList = DataBox.get(inputs, "input", ListDataType.GENERIC_LIST).orElse(List.of());
+        if (inputList.isEmpty()) {
+            return Map.of(
+                "first", CompletableFuture.completedFuture(DataType.ANY.create(Optional.empty())),
+                "rest", CompletableFuture.completedFuture(ListDataType.GENERIC_LIST.create(List.of()))
+            );
+        }
+        var shouldExtractHead = DataBox.get(settings, "extraction_mode",
+            EXTRACTION_MODE_TYPE).orElse(ExtractionMode.EXTRACT_HEAD) == ExtractionMode.EXTRACT_HEAD;
+        var lastIndex = inputList.size() - 1;
+        var extractionIndex = shouldExtractHead ? 0 : lastIndex;
+        var subListOffset = shouldExtractHead ? 1 : 0;
+        var outputType = this.findExactDatatype(inputTypes.get("input"));
+        return Map.of(
+            "first", CompletableFuture.completedFuture(outputType.create(
+                inputList.get(extractionIndex))),
+            "rest", CompletableFuture.completedFuture(outputType.listOf().create(
+                (List<Object>) inputList.subList(subListOffset, lastIndex + subListOffset)))
+        );
+    }
+
+    @SuppressWarnings({"unchecked", "rawtypes"})
+    private <T> DataType<T> findExactDatatype(DataType type) {
+        if (type.equals(DataType.NUMBER)) {
+            type = DataType.NUMBER;
+        } else if (type.equals(DataType.STRING)) {
+            type = DataType.STRING;
+        } else if (type.equals(DataType.BOOLEAN)) {
+            type = DataType.BOOLEAN;
+        } else if (type.equals(DataType.JSON_OBJECT)) {
+            type = DataType.JSON_OBJECT;
+        } else {
+            type = DataType.ANY;
+        }
+        return type;
+    }
```

Comment:
? the input `type` here is always a list type, so the output here will always be `DataType.ANY`.

This method should look something like:
```
return type instanceof ListDataType<?> list
    ? list.contains()
    : DataType.ANY;
```

Also, you can use `DataType<?>` instead of `DataType` and then just `return` in each branch to remove one of the suppressed warnings, I think.

Reply:
removed raw usages of types